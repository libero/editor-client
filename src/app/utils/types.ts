import React from 'react';

export type ReactFCProps<FC> = FC extends React.FC<infer P> ? P : {};

export type ValueOf<T> = T[keyof T];
