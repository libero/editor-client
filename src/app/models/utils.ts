import { pick, get } from 'lodash';
import { Schema, SchemaSpec } from 'prosemirror-model';

import { nodes } from 'app/models/config/nodes';
import { marks } from 'app/models/config/marks';

export function makeSchemaFromConfig(topNode: string, nodeNames: string[], markNames: string[]): Schema {
  const filteredNodes = pick(nodes, nodeNames);
  const filteredMarks = pick(marks, markNames);

  return new Schema({
    nodes: filteredNodes,
    marks: filteredMarks,
    topNode
  } as SchemaSpec<keyof typeof filteredNodes, keyof typeof filteredMarks>);
}

export function getTextContentFromPath(el: ParentNode, path): string {
  return get(el.querySelector(path), 'textContent');
}
