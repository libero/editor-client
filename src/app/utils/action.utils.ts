import { createAction, SimpleActionCreator } from 'redux-act';

export interface Action<T> {
  type: string;
  payload: T;
}

export interface AsyncAction<R, T> {
  request: SimpleActionCreator<R>;
  success: SimpleActionCreator<T>;
  error: SimpleActionCreator<Error>;
}

export function createAsyncAction<R, T>(type: string): AsyncAction<R, T> {
  return {
    request: createAction<R>(type + '_REQUEST'),
    success: createAction<T>(type + '_SUCCESS'),
    error: createAction<Error>(type + '_ERROR')
  } as AsyncAction<R, T>;
}
