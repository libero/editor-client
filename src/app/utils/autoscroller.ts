const WINDOW_SCROLL_ZONE = 300;

interface Coords {
  x: number;
  y: number;
}

interface ScrollUpdateInterface {
  translate: Coords;
  maxTranslate: Coords;
  minTranslate: Coords;
  minScroll: Coords;
  width: number;
  height: number;
}

export class AutoScroller {
  private static instance: AutoScroller;

  public static getInstance(): AutoScroller {
    if (!AutoScroller.instance) {
      AutoScroller.instance = new AutoScroller();
    }
    return AutoScroller.instance;
  }

  private interval: number | null;
  private container: Element;
  private isAutoScrolling: boolean;

  constructor() {
    this.container = window.document.scrollingElement || window.document.documentElement;
  }

  public clear(): void {
    if (this.interval == null) {
      return;
    }

    clearInterval(this.interval);
    this.interval = null;
  }

  public updateScroll(event: DragEvent, dropContainer: Element): void {
    const containerRect = this.container.getBoundingClientRect();
    const dropRect = dropContainer.getBoundingClientRect();

    const minTranslate: Coords = { x: 0, y: 0 };
    const maxTranslate: Coords = {
      x: containerRect.width,
      y: containerRect.height
    };

    const minScroll = {
      x: dropRect.left + this.container.scrollLeft,
      y: dropRect.top + this.container.scrollTop
    };

    this.doAutoscroll({
      translate: { x: event.clientX, y: event.clientY },
      minTranslate,
      maxTranslate,
      minScroll,
      width: WINDOW_SCROLL_ZONE * 2,
      height: WINDOW_SCROLL_ZONE * 2
    });
  }

  private doAutoscroll({ translate, minTranslate, maxTranslate, width, height }: ScrollUpdateInterface): void {
    const direction = {
      x: 0,
      y: 0
    };

    const speed = {
      x: 1,
      y: 1
    };

    const acceleration = {
      x: 10,
      y: 10
    };

    const { scrollTop, scrollLeft, scrollHeight, scrollWidth, clientHeight, clientWidth } = this.container;

    const isTop = scrollTop === 0;
    const isBottom = scrollHeight - scrollTop - clientHeight === 0;
    const isLeft = scrollLeft === 0;
    const isRight = scrollWidth - scrollLeft - clientWidth === 0;

    if (translate.y >= maxTranslate.y - height / 2 && !isBottom) {
      // Scroll Down
      direction.y = 1;
      speed.y = acceleration.y * Math.abs((maxTranslate.y - height / 2 - translate.y) / height);
    } else if (translate.x >= maxTranslate.x - width / 2 && !isRight) {
      // Scroll Right
      direction.x = 1;
      speed.x = acceleration.x * Math.abs((maxTranslate.x - width / 2 - translate.x) / width);
    } else if (translate.y <= minTranslate.y + height / 2 && !isTop) {
      // Scroll Up
      direction.y = -1;
      speed.y = acceleration.y * Math.abs((translate.y - height / 2 - minTranslate.y) / height);
    } else if (translate.x <= minTranslate.x + width / 2 && !isLeft) {
      // Scroll Left
      direction.x = -1;
      speed.x = acceleration.x * Math.abs((translate.x - width / 2 - minTranslate.x) / width);
    }

    if (this.interval) {
      this.clear();
      this.isAutoScrolling = false;
    }

    if (direction.x !== 0 || direction.y !== 0) {
      this.interval = window.setInterval(() => {
        this.isAutoScrolling = true;
        const offset = {
          left: speed.x * direction.x,
          top: speed.y * direction.y
        };
        this.container.scrollTop += offset.top;
        this.container.scrollLeft += offset.left;
      }, 5);
    }
  }
}
