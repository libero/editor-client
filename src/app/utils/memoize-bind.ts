import { memoize } from 'lodash';
import objectHash from 'object-hash';

export const memoizeBind = memoize(
  (fn: Function, ...args: any[]) => {
    return fn.bind(null, ...args);
  },
  (fn: Function, firstArg: any) => {
    return objectHash({
      fn,
      firstArg
    });
  }
);
