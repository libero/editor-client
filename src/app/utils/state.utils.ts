import { EditorState } from 'prosemirror-state';
import { cloneDeepWith } from 'lodash';

import { Manuscript, ManuscriptDiff } from 'app/types/manuscript';

export interface LoadableState<T> {
  data: T | undefined;
  isLoading: boolean;
  error: Error | undefined;
}

export function getInitialLoadableState<T>(): LoadableState<T> {
  return {
    data: undefined,
    isLoading: false,
    error: undefined
  };
}

export function getLoadableStateProgress<T>(): LoadableState<T> {
  return {
    data: undefined,
    isLoading: true,
    error: undefined
  };
}

export function getLoadableStateSuccess<T>(data: T): LoadableState<T> {
  return {
    data,
    isLoading: false,
    error: undefined
  };
}

export function getLoadableStateError<T>(error: Error): LoadableState<T> {
  return {
    data: undefined,
    isLoading: false,
    error
  };
}

export interface ManuscriptHistory {
  past: ManuscriptDiff[];
  present: Manuscript;
  future: ManuscriptDiff[];
}

export function getInitialHistory(present: Manuscript): ManuscriptHistory {
  return { past: [], present, future: [] };
}

export function cloneManuscript(manuscript: Manuscript): Manuscript {
  const cloneCustomizer = (value): EditorState | undefined => (value instanceof EditorState ? value : undefined);
  return cloneDeepWith(manuscript, cloneCustomizer);
}
