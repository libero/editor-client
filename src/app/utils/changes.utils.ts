import { cloneDeepWith } from 'lodash';
import { EditorState } from 'prosemirror-state';

import { JSONObject } from 'app/types/utility.types';
import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { BatchChange } from 'app/utils/history/batch-change';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';
import { RearrangingChange } from 'app/utils/history/rearranging-change';
import { UpdateObjectChange } from 'app/utils/history/update-object-change';
import { DeleteObjectChange } from 'app/utils/history/delete-object-change';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { Affiliation } from 'app/models/affiliation';
import { Person } from 'app/models/person';
import { Reference } from 'app/models/reference';
import { RelatedArticle } from 'app/models/related-article';
import { Keyword } from 'app/models/keyword';
import { AddObjectChange } from 'app/utils/history/add-object-change';

export function manuscriptEntityToJson<T>(object: T): JSONObject {
  return cloneDeepWith(object, (value) => {
    if (value instanceof EditorState) {
      return value.toJSON();
    }
  });
}

export function deserializeChanges(changesJson: JSONObject[]): Change[] {
  return changesJson.map((changeData) => {
    switch (changeData.type) {
      case 'prosemirror':
        return ProsemirrorChange.fromJSON(changeData);

      case 'rearranging':
        return RearrangingChange.fromJSON(changeData);

      case 'batch':
        return BatchChange.fromJSON(changeData);

      case 'update-object':
        return UpdateObjectChange.fromJSON(changeData);

      case 'delete-object':
        return DeleteObjectChange.fromJSON(changeData);

      case 'add-object':
        return AddObjectChange.fromJSON(changeData);

      default:
        throw new TypeError(
          `Cannot deserialize ${changeData.type} change. Change type invalid, JSON object malformed or handling not supported.`
        );
    }
  });
}

export function deserializeBackmatter(path: string, json: JSONObject): BackmatterEntity {
  if (path.indexOf('affiliations') >= 0) {
    return new Affiliation(json);
  }

  if (path.indexOf('authors') >= 0) {
    return new Person(json);
  }

  if (path.indexOf('references') >= 0) {
    return new Reference(json);
  }

  if (path.indexOf('relatedArticles') >= 0) {
    return new RelatedArticle(json);
  }

  if (path.match(/keywordGroups\.[^.]+\.keywords/)) {
    return new Keyword(json);
  }

  throw new Error(`deserialization of backmatter entity for  ${path} is not implemented or provided path is invalid`);
}

export function applyChangesFromServer(manuscript: Manuscript, changesJson: JSONObject[]): Manuscript {
  const allChanges = new BatchChange(deserializeChanges(changesJson));
  return allChanges.applyChange(manuscript);
}
