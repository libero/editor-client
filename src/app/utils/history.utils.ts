import { Transaction } from 'prosemirror-state';

import { ManuscriptHistory } from './state.utils';
import { ManuscriptDiff, ManuscriptDiffValues } from 'app/types/manuscript';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';

export function updateManuscriptState(
  state: ManuscriptHistory,
  propName: string,
  transaction: Transaction
): ManuscriptHistory {
  const change = new ProsemirrorChange(propName, transaction);

  if (transaction.docChanged) {
    return {
      past: [...state.past, change],
      present: change.applyChange(state.present),
      future: []
    } as ManuscriptHistory;
  } else {
    return {
      ...state,
      present: change.applyChange(state.present)
    };
  }
}

export function undoChange(state: ManuscriptHistory): ManuscriptHistory {
  const past = [...state.past];
  const change = past.pop();

  return {
    past,
    present: change.rollbackChange(state.present),
    future: [change, ...state.future]
  };
}

export function createDiff(changes: Record<string, ManuscriptDiffValues>): ManuscriptDiff {
  return {
    ...changes,
    _timestamp: Date.now()
  } as ManuscriptDiff;
}

export function redoChange(state: ManuscriptHistory): ManuscriptHistory {
  const future = [...state.future];
  const change = future.shift();

  return {
    past: [...state.past, change],
    present: change.applyChange(state.present),
    future: future
  };
}
