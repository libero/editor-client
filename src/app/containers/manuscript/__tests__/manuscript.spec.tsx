import React from 'react';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Backdrop, CircularProgress, Hidden, Container } from '@material-ui/core';

import { getInitialHistory, getLoadableStateProgress, getLoadableStateSuccess } from '../../../utils/state.utils';
import { ManuscriptContainer } from '../index';
import { EditorState } from 'prosemirror-state';

jest.mock('@material-ui/core');

describe('Manuscript container', () => {
  const mockStore = configureMockStore([]);

  beforeEach(() => {
    (Hidden as jest.Mock).mockImplementation(({ children }) => <div data-cmp="hidden">{children}</div>);
    (Container['render'] as jest.Mock).mockImplementation(({ children }) => <div data-cmp="container">{children}</div>);
  });

  afterEach(() => {
    (Hidden as jest.Mock).mockReset();
    (Container['render'] as jest.Mock).mockReset();
  });

  it('renders when data is loaded', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      keywordGroups: {},
      authors: []
    });
    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });
    const wrapper = create(
      <Provider store={store}>
        <ManuscriptContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading spinner when manuscript is loading', () => {
    (Backdrop['render'] as jest.Mock).mockImplementationOnce(({ children }) => (
      <div data-cmp="backdrop">{children}</div>
    ));
    (CircularProgress['render'] as jest.Mock).mockImplementationOnce(() => <div data-cmp="circular-progress"></div>);

    const store = mockStore({ manuscript: getLoadableStateProgress() });
    const wrapper = create(
      <Provider store={store}>
        <ManuscriptContainer />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();

    (Backdrop['render'] as jest.Mock).mockReset();
    (CircularProgress['render'] as jest.Mock).mockReset();
  });
});
