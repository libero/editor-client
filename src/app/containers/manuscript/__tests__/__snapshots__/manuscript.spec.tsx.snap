// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manuscript container renders loading spinner when manuscript is loading 1`] = `
<div
  data-cmp="backdrop"
>
  <div
    data-cmp="circular-progress"
  />
</div>
`;

exports[`Manuscript container renders when data is loaded 1`] = `
Array [
  <nav
    aria-label="mailbox folders"
    className="makeStyles-drawer-1"
  >
    <div
      data-cmp="hidden"
    />
    <div
      data-cmp="hidden"
    />
  </nav>,
  <div
    className="makeStyles-contentWrapper-13"
    onClick={[Function]}
  >
    <div
      aria-hidden="true"
      className="makeStyles-toolbarPlaceholder-16"
    />
    <div
      className="makeStyles-content-14"
      onClick={[Function]}
    >
      <div
        className="makeStyles-root-17 makeStyles-plainVariant-19"
      >
        <label
          className="makeStyles-label-20"
        >
          Title
        </label>
        <div
          className="prosemirrorContainer"
        />
      </div>
      <div
        onClick={[Function]}
      >
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <section>
          <div
            className="makeStyles-root-17 makeStyles-plainVariant-19"
          >
            <label
              className="makeStyles-label-20"
            >
              Authors
            </label>
            <div
              className="makeStyles-sortableContainer-22"
            >
               
               
            </div>
          </div>
        </section>
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <section>
          <div
            className="makeStyles-root-17 makeStyles-plainVariant-19"
          >
            <label
              className="makeStyles-label-20"
            >
              Affiliations
            </label>
          </div>
        </section>
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
      </div>
      <div
        className="makeStyles-root-17 makeStyles-plainVariant-19"
      >
        <label
          className="makeStyles-label-20"
        >
          Abstract
        </label>
        <div
          className="prosemirrorContainer"
        />
      </div>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-15"
        onClick={[Function]}
      />
      <div
        className="makeStyles-root-17 makeStyles-plainVariant-19"
      >
        <label
          className="makeStyles-label-20"
        >
          Impact statement
        </label>
        <div
          className="prosemirrorContainer"
        />
      </div>
      <div
        aria-hidden="true"
        className="makeStyles-spacer-15"
        onClick={[Function]}
      />
      <div
        onClick={[Function]}
      >
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <div
          className="makeStyles-sectionContainer-37 makeStyles-root-17 makeStyles-plainVariant-19"
        >
          <label
            className="makeStyles-label-20"
          >
            Article information
          </label>
          <section
            className="makeStyles-infoSection-38"
          >
            <div>
              <strong>
                 Article DOI: 
              </strong>
               
              
            </div>
          </section>
        </div>
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <section>
          <div
            className="makeStyles-root-17 makeStyles-plainVariant-19"
          >
            <label
              className="makeStyles-label-20"
            >
              Related articles
            </label>
            <ul
              className="makeStyles-list-41"
            />
          </div>
        </section>
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <section>
          <div
            className="makeStyles-root-17 makeStyles-plainVariant-19"
          >
            <label
              className="makeStyles-label-20"
            >
              Author Information
            </label>
          </div>
        </section>
        <div
          aria-hidden="true"
          className="makeStyles-spacer-15"
          onClick={[Function]}
        />
        <section>
          <div
            className="makeStyles-root-17 makeStyles-plainVariant-19"
          >
            <label
              className="makeStyles-label-20"
            >
              References
            </label>
            <ul
              className="makeStyles-list-46"
            />
          </div>
        </section>
      </div>
    </div>
  </div>,
]
`;
